tests/components/matrix/test_login.py

```diff
@@ -0,0 +1,121 @@
+from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
+
+from tests.components.matrix.conftest import (
+    TEST_DEVICE_ID,
+    TEST_MXID,
+    TEST_PASSWORD,
+    TEST_TOKEN,
+)
+
+
+@dataclass
+class LoginTestParameters:
+    """Dataclass of parameters representing the login parameters and expected result state."""
+
+    valid_password: bool
+    access_token: dict[str, str]
+    expected_login_state: bool
+    expected_caplog_messages: set[str]
+    expected_expection: type(Exception) | None = None
+
+
+good_password_missing_token = LoginTestParameters(
+    password=TEST_PASSWORD,
+    access_token={},
+    expected_login_state=True,
+    expected_caplog_messages={"Logging in using password"},
+)
+
+good_password_bad_token = LoginTestParameters(
+    password=TEST_PASSWORD,
+    access_token={TEST_MXID: "WrongToken"},
+    expected_login_state=True,
+    expected_caplog_messages={
+        "Restoring login from stored access token",
+        "Restoring login from access token failed: M_UNKNOWN_TOKEN, Invalid access token passed.",
+        "Logging in using password",
+    },
+)
+
+bad_password_good_access_token = LoginTestParameters(
+    password="WrongPassword",
+    access_token={TEST_MXID: TEST_TOKEN},
+    expected_login_state=True,
+    expected_caplog_messages={
+        "Restoring login from stored access token",
+        f"Successfully restored login from access token: user_id '{TEST_MXID}', device_id '{TEST_DEVICE_ID}'",
+    },
+)
+
+bad_password_bad_access_token = LoginTestParameters(
+    password="WrongPassword",
+    access_token={TEST_MXID: "WrongToken"},
+    expected_login_state=False,
+    expected_caplog_messages={
+        "Restoring login from stored access token",
+        "Restoring login from access token failed: M_UNKNOWN_TOKEN, Invalid access token passed.",
+        "Logging in using password",
+        "Login by password failed: status_code, LoginError",
+    },
+    expected_expection=ConfigEntryAuthFailed,
+)
+
+bad_password_missing_access_token = LoginTestParameters(
+    password="WrongPassword",
+    access_token={},
+    expected_login_state=False,
+    expected_caplog_messages={
+        "Logging in using password",
+        "Login by password failed: status_code, LoginError",
+    },
+    expected_expection=ConfigEntryAuthFailed,
+)
+
+
+@pytest.mark.parametrize(
+    "params",
+    [
+        good_password_missing_token,
+        good_password_bad_token,
+        bad_password_good_access_token,
+        bad_password_bad_access_token,
+        bad_password_missing_access_token,
+    ],
+)
+async def test_login(
+    matrix_bot: MatrixBot, caplog: pytest.LogCaptureFixture, params: LoginTestParameters
+):
+    """Test logging in with the given parameters and expected state."""
+    await matrix_bot._client.logout()
+    password = "WrongPassword"
+    if params.valid_password:
+        password = TEST_PASSWORD
+    matrix_bot._password = password
```

Comment:
```suggestion
    matrix_bot._password = params.password
```

From your draft I saw you had even a better idea :)